app/controllers/automated_tests_controller.rb

```diff
@@ -66,23 +66,29 @@ def student_interface
       @revision  = repo.get_latest_revision
       @revision_number = @revision.revision_number
 
-      @test_script_results = TestScriptResult.find_by_grouping_id(@grouping.id)
+      @test_script_results = TestResult.where(grouping:
+        @grouping).order(created_at: :desc)
 
-      @token = Token.find_by_grouping_id(@grouping.id)
-      if @token
-        @token.reassign_tokens_if_new_day()
-      end
+      @token = fetch_latest_tokens_for_grouping(@grouping)
+    end
+  end
+
+  def execute_test_run
+    @assignment = Assignment.find(params[:id])
+    @student = current_user
+    @grouping = @student.accepted_grouping_for(@assignment.id)
+    @token = fetch_latest_tokens_for_grouping(@grouping)
 
-      # For running tests
-      if params[:run_tests] && ((@token && @token.tokens > 0) || @assignment.unlimited_tokens)
-        result = run_tests(@grouping.id)
-        if result == nil
-          flash[:notice] = I18n.t("automated_tests.tests_running")
-        else
-          flash[:failure] = result
-        end
+    # For running tests
+    if ((@token && @token.tokens > 0) || @assignment.unlimited_tokens)
+      result = run_tests(@grouping.id)
+      if result == nil
+        flash[:notice] = I18n.t("automated_tests.tests_running")
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.